### PR TITLE
Zowe copybook tracker #208

### PR DESCRIPTION
**Pull request by zacanbrcom**
Address the issue #180 and extends the #141.
This POC consider a very basic scenario and didn't include a proper definition of setting.json that may be introduced later on.

End user scenario

Cobol file opened with a missing copybook
the related dependency file is generated
The copybook tracker extension ask user credential to connect to mf and download the copybook into the .copybooks folder
The LSP trigger a new analysis for the cobol file and the syntax error of copybook not found disappear.

### CODE REVIEW

File: clients/cobol-lsp-vscode-extension/tslint.json [hunk 1]
```diff
@@ -31,7 +31,8 @@
         "arrow-parens": [
             true,
             "ban-single-arg-parens"
-        ]
+        ],
+        "prettier": true
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/tslint.json]: Why do we store tslint config on the repository?

File: clients/cobol-lsp-vscode-extension/src/CopybooksDownloader.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+const SETTINGS_ROOT = "cobol-language-support";
+
+export class CopybooksDownloader {
+    public constructor(private zoweApi: ZoweApi) { }
+
+    /**
+     * @param copybooks array of copybooks names to download
+     */
+    // tslint:disable-next-line: cognitive-complexity
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/src/CopybooksDownloader.ts]: 😠

File: clients/cobol-lsp-vscode-extension/tslint.json (hunk 1 above)
Comment (ishche) [reply clients/cobol-lsp-vscode-extension/tslint.json]: To have it nice for everyone.

Comment (temanbrcom) [reply clients/cobol-lsp-vscode-extension/tslint.json]: Isn't tslint a dead project?

Comment (ishche) [reply clients/cobol-lsp-vscode-extension/tslint.json]: It is. But I believe that eslint understand tslint configuration.

Comment (temanbrcom) [reply clients/cobol-lsp-vscode-extension/tslint.json]: I think that it deserves an additional user story to provide a proper eslint config for our requirements.

Comment (ishche) [reply clients/cobol-lsp-vscode-extension/tslint.json]: Also, the story should be about switching to eslint. I agree.

**Comment by temanbrcom** [comment]
What is the status of this PR? Should it be merged?

File: clients/cobol-lsp-vscode-extension/package.json [hunk 3]
```diff
@@ -38,7 +50,24 @@
...
-        ]
+        ],
+        "configuration": {
+            "title": "Broadcom COBOL LSP",
+            "properties": {
+                "broadcom-cobol-lsp.cpy-manager.profiles": {
+                    "type": "Object",
+                    "description": "Program name to profile map"
+                },
+                "broadcom-cobol-lsp.cpy-manager.paths": {
```
Comment (sergiuilie) [inline clients/cobol-lsp-vscode-extension/package.json]: in order to use autocomplete in settings.json we can not combine different properties under one structure to have something like: 
`broadcom-cobol-lsp.cpy-manager: {
                                                           paths: {},
                                                           profiles: {}
}
`
this [link](https://json-schema.org/understanding-json-schema/reference/index.html) is useful for future implementation

File: clients/cobol-lsp-vscode-extension/src/ProfileService.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            } else {
+                items.push(item);
+            }
+        });
+
+        const selectedProfile = await vscode.window.showQuickPick(items,
+            { placeHolder: defaultName, canPickMany: false });
+        if (selectedProfile) {
+            let settingsProfiles = vscode.workspace.getConfiguration().get(SETTINGS_SECTION + ".profiles");
+            settingsProfiles = settingsProfiles ? settingsProfiles : {};
```
Comment (sergiuilie) [inline clients/cobol-lsp-vscode-extension/src/ProfileService.ts]: you don't get any error here? can you assign empty object to the settingsProfile?

File: clients/cobol-lsp-vscode-extension/src/ZoweApi.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        // default should be https
+        const rpath = `/zosmf/restfiles/ds/${dataset}(${member})`;
+        return await RestClient.getExpectString(session, rpath, [{
+            "Content-Type": "application/json", "X-CSRF-ZOSMF-HEADER": "" }]);
+    }
+
+    public async listMembers(dataset: string, profileName: string): Promise<string[]> {
+        // default should be https
+        const session: Session = await this.createSession(profileName);
+        const rpath = `/zosmf/restfiles/ds/${dataset}/member`;
```
Comment (sergiuilie) [inline clients/cobol-lsp-vscode-extension/src/ZoweApi.ts]: we are depending a lot on this hardcoded parts from Zowe

File: com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/TestModule.java [hunk 6]
```diff
@@ -16,6 +16,8 @@
 package com.ca.lsp.cobol;
 
 import com.broadcom.lsp.cdi.module.DefaultModule;
+import com.broadcom.lsp.domain.cobol.databus.impl.AbstractDataBusBroker;
```
Comment (sergiuilie) [inline com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/TestModule.java]: what does it have to do with the client?

File: clients/cobol-lsp-vscode-extension/package.json [hunk 7]
```diff
@@ -38,7 +50,24 @@
                     ".cob"
                 ]
             }
-        ]
+        ],
+        "configuration": {
+            "title": "Broadcom COBOL LSP",
+            "properties": {
+                "broadcom-cobol-lsp.cpy-manager.profiles": {
```
Comment (sergiuilie) [inline clients/cobol-lsp-vscode-extension/package.json]: can the profile be a string?

File: clients/cobol-lsp-vscode-extension/src/ProfileService.ts (hunk 4 above)
Comment (ishche) [reply clients/cobol-lsp-vscode-extension/src/ProfileService.ts]: It is an empty object according to package.json
I use it as a map to store a profile name for program.

File: clients/cobol-lsp-vscode-extension/src/ZoweApi.ts (hunk 5 above)
Comment (ishche) [reply clients/cobol-lsp-vscode-extension/src/ZoweApi.ts]: It's not Zowe, it's zOSMF. Can you describe more, what problem do you see here?

File: clients/cobol-lsp-vscode-extension/package.json (hunk 7 above)
Comment (ishche) [reply clients/cobol-lsp-vscode-extension/package.json]: Zowe Profile is a pretty complex structure, here I use an object to store relation with program name (string) to profile name (string)

File: clients/cobol-lsp-vscode-extension/src/CopybooksDownloader.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        const copybooks: string[] = fs.readFileSync(uri.fsPath).toString().split("\n")
+            .filter(e => e.trim().length > 0)
+            .map(e => e.trim());
+        const missingCopybooks: string[] = this.readMissingCopybooks(missingCopybooksFilePath);
+        if (vscode.workspace.workspaceFolders.length === 0) {
+            vscode.window.showErrorMessage("No workspace folder opened.");
+            return;
+        }
+
+        const cb: Set<string> = new Set(copybooks);
```
[the remark this answers is not in the document]
Comment (ishche) [reply clients/cobol-lsp-vscode-extension/src/CopybooksDownloader.ts]: ds seem pretty clear for me, cb - not)
renamed

File: clients/cobol-lsp-vscode-extension/src/ZoweApi.ts (hunk 5 above)
Comment (sergiuilie) [reply clients/cobol-lsp-vscode-extension/src/ZoweApi.ts]: it is not a problem, the thing is that we hardcode the path and if they change something we are dependent on them

**Comment by zacanbrcom** [comment]
@shciubrcom after address the conflict file can we merge it or u planned some enhancement that are in progress?

**Comment by sergiuilie** [comment]
> @shciubrcom after address the conflict file can we merge it or u planned some enhancement that are in progress?

maybe to fix the che problem, first?

**Comment by ishche** [comment]
> > @shciubrcom after address the conflict file can we merge it or u planned some enhancement that are in progress?
> 
> maybe to fix the che problem, first?

Testing still in progress

File: clients/cobol-lsp-vscode-extension/src/ZoweApi.ts (hunk 5 above)
Comment (temanbrcom) [reply clients/cobol-lsp-vscode-extension/src/ZoweApi.ts]: I agree I would prefer to have these values in some resource file

**Pull request merged**